### DISCUSSION TITLE
Utilizing Block Restriction To Generate Two GlobalStrain Objects

### DISCUSSION THREAD

**Discussion by lkuna24** [idaholab/moose#27885] [Q&A Modules: Solid mechanics]
Hi All,
I am trying to carry out a simulation of a thin film with an applied epitaxy in the form of an eigenstrain.  In the simulation the thin film sits on top of a substrate which imposes a uniform mismatch strain throughout the film (so long as the film is sufficiently thin).  The domain (both film and substrate) is periodic in the x and y direction.  Applying GlobalStrain to both the film and substrate with just the epitaxy (in the form of an eigenstrain) acting on the film yields some unexpected results, but block restricting and applying GlobalStrain to only the film with the epitaxy applied gives the expected result.  I wanted to attempt to separate out GlobalStrain utilizing block restriction as shown in the input file below, but I get the error:
```
*** ERROR ***
The following error occurred in the Problem 'MOOSE Problem' of type FEProblem.

Material property 'global_strain_film', requested by 'Ferroelectric_SM_strain' is not defined on block 0
Material property 'global_strain_sub', requested by 'Substrate_SM_strain' is not defined on block 1
```

As I need the whole domain to be periodic, I believe I need a GlobalStrain on both blocks.  According to the documentation here [GlobalStrain](https://mooseframework.inl.gov/syntax/Modules/TensorMechanics/GlobalStrain/), it seems as though I should be able to separate GlobalStrain Objects utilizing block restriction, or is this actually not possible?  

Ultimately there will be another eigenstrain contributing to the thin film which will be coupled to the evolution of polarization throughout the volume of the film.  But for now, I am just trying to see if I can establish the mechanical properties of the film and substrate.

Thanks for you time.

```
[Mesh]
  [gen]
    type = GeneratedMeshGenerator
    dim = 3
    nx = 10
    ny = 10
    nz = 20
    xmin = -10.0
    xmax = 10.0
    ymin = -10.0
    ymax = 10.0
    zmin = -10.0
    zmax = 10.0
    elem_type = HEX8
  []
  [./cnode]
    input = gen
    type = ExtraNodesetGenerator
    coord = '-10.0 -10.0 -10.0'
    new_boundary = 100
  [../]

  [subdomains]
    type = SubdomainBoundingBoxGenerator
    input = cnode
    bottom_left = '-10.0 -10.0 -10.0'
    block_id = 1
    top_right = '10.0 10.0 0.0'
    location = INSIDE
  []
  [film_interface]
    type = SideSetsBetweenSubdomainsGenerator
    input = subdomains
    primary_block = 0
    paired_block = 1
    new_boundary = 52
  []
[]

[GlobalParams]
  displacements = 'u_x u_y u_z'
[]

[Variables]
  [./global_strain_film]
    order = SIXTH
    family = SCALAR
    block = '0'
  [../]

  [./global_strain_substrate]
    order = SIXTH
    family = SCALAR
    block = '1'
  [../]

  [./u_x]
    order = FIRST
    family = LAGRANGE
    block = '0 1'
  [../]
  [./u_y]
    order = FIRST
    family = LAGRANGE
    block = '1 0'
  [../]
  [./u_z]
    order = FIRST
    family = LAGRANGE
    block = '1 0'
  [../]
[]


[Materials]

  [./eigen_strain]
    type = ComputeEigenstrain
    #    eigen_base = 'x y z yz xz xy'
    eigen_base = '0.005 0.005 0.003725 0 0 0'
    eigenstrain_name = epitaxy
    block = '1 0'
  [../]

  [./elasticity_tensor_1]
    type = ComputeElasticityTensor
    fill_method = symmetric9
    C_ijkl = '179.073 66.71 66.71 179.073 66.71 179.073 82.6446 82.6446 82.6446'
  [../]

  [./stress_1]
    type = ComputeLinearElasticStress
  [../]
[]

[Physics]
  [SolidMechanics]
    [QuasiStatic]
      [./Ferroelectric_SM]
        strain = SMALL
        add_variables = true
        incremental = false
        global_strain = global_strain_film
        eigenstrain_names = 'epitaxy'
        generate_output = 'strain_xx strain_xy strain_xz strain_yy strain_yz strain_zz vonmises_stress'
	       block = '0'
      [../]

      [./Substrate_SM]
        strain = SMALL
        add_variables = true
        global_strain = global_strain_substrate
  #      eigenstrain_names = 'epitaxy'
        incremental = false
        generate_output = 'strain_xx strain_xy strain_xz strain_yy strain_yz strain_zz vonmises_stress'
        block = '1'
      [../]
    []

    [./GlobalStrain]
      [./global_strain_film]
        scalar_global_strain = global_strain_film
        displacements = 'u_x u_y u_z'
        auxiliary_displacements = 'disp_x disp_y disp_z'
        global_displacements = 'ug_x ug_y ug_z'
        block = '0'
      [../]
      [./global_strain_sub]
        scalar_global_strain = global_strain_substrate
        displacements = 'u_x u_y u_z'
        auxiliary_displacements = 'disps_x disps_y disps_z'
        global_displacements = 'ugs_x ugs_y ugs_z'
        block = '1'
      [../]
    [../]
  []
[]

[BCs]
  [./Periodic]
    [./xy]
      auto_direction = 'x y'
      variable = 'u_x u_y u_z'
    [../]
  [../]
  [./centerfix_x]
    type = DirichletBC
    boundary = 100
    variable = u_x
    value = 0
  [../]
  [./centerfix_y]
    type = DirichletBC
    boundary = 100
    variable = u_y
    value = 0
  [../]
  [./centerfix_z]
    type = DirichletBC
    boundary = 100
    variable = u_z
    value = 0
  [../]
[]

[Preconditioning]
  [./smp]
    type = SMP
    full = true
    petsc_options = '-snes_ksp_ew'
    petsc_options_iname = '-ksp_gmres_restart -snes_atol -snes_rtol -ksp_rtol -pc_type'
    petsc_options_value = '    160             1e-8        1e-6      1e-5        bjacobi'
  [../]
[]

[Executioner]
  type = Transient
  solve_type = 'PJFNK'
  scheme = 'implicit-euler'
  dtmin = 1e-13
  dtmax = 2.0
  end_time = 10
  l_max_its = 200
  automatic_scaling = true
  [./TimeStepper]
    type = IterationAdaptiveDT
    optimal_iterations = 8
    growth_factor = 1.1
    cutback_factor = 0.8
    linear_iteration_ratio = 1000
    dt = 0.5
  [../]
  verbose = true
  nl_max_its = 16
[]

[Outputs]
  print_linear_residuals = false
  perf_graph = false
  [./out]
    type = Exodus
    file_base = out_GS_Test
    interval = 2
  [../]
[]
```

**Comment by GiudGiud** [comment]
@SudiptaBiswas 

Block restriction seems to be implemented so this looks like the expected syntax
Use Debug/show_material_props to see what names are being used for the global strain material property when they are being declared. Then adapt the input to use the correct names

**Comment by mangerij** [comment]
@lkuna24 your zz epitaxy should be minus :)

I just want to (re)iterate that this is part of a long standing issue within Ferret of computing ferroelectric thin film phase diagrams:

See 
Appl. Phys. Lett. 78, 3878–3880 (2001)
https://doi.org/10.1063/1.1377855

We get many requests (approx 1-2/month) to be able to solve this problem and I think its time we finally pin this down.

We are slowly refactoring things to make them use more MOOSE systems but we still can't find what is going on. As soon as we throw in a second eigenstrain (proportional to the square of the phase field variable), we find that it dominates the calculation whereas instead we expect that at the end of the solve, the global strain of the film block should be very much equal to the eigenstrain "epitaxy" that we apply above. 

As such, the epitaxial condition (eigenstrain) coupling to the phase field problem is only a weak influence.

**Reply by GiudGiud** [reply]
hi @mangerij 
maybe @bwspenc will know who to bring in on this
